core: kernel logging

OlegHahm opened the pull request:
Introduces `LOG_KERNEL` for printing kernel messages (similar to `printk()` in Linux). This allows to set the default logging level to a very silent level, in order to make better use of the logging system throughout the system.

kaspar030 commented:
- ```printk``` is just in-kernel printf.

- Why should the kernel not adhere to other log levels? "Hello, this is RIOT!" should be ```LOG_WARNING```.

OlegHahm commented:
Yeah, I know. I just wanted to find a nice way to have "Hello, this is RIOT!" being always printed, even if the debug level is set low.

kaspar030 commented:
;)

Why not just always "puts" it?

OlegHahm commented:
Hm, okay, replace "always" with "almost always" in my former statement. I think it is a nice idea to mute the node completely easily.

kaspar030 commented:
```LOG_CRITICAL```? Seriously, a new log-level for this?

OlegHahm commented:
I think "Hello, this is RIOT!" doesn't look very critical.

OlegHahm commented:
How about this?

kaspar030 commented:
Yes! RIOT has printk now!

```#define LOG_IMPORTANT LOG_CRITICAL```?

OlegHahm commented:
> #define LOG_IMPORTANT LOG_CRITICAL?

What? Why? There's nothing like `LOG_IMPORTANT` in the syslog POSIX standard.

OlegHahm commented:
Addressed comments from #3506.

OlegHahm commented:
btw. any idea why it doesn't work as intended with the former enum solution? Performing a `#if (LOG_LEVEL > LOG_LEVEL_FOO)` check works apparently only with defines.

kaspar030 commented:
> Performing a #if (LOG_LEVEL > LOG_LEVEL_FOO) check works apparently only with defines.

Ah, the preprocessor doesn't work with enum values.

OlegHahm commented:
@kaspar030, what's your opinion on this?

kaspar030 commented:
25 lines in order to maybe silence two printfs. I'd say, no extra kernel muting needed, let's either say, those lines belong to RIOT and we always print them, or just make them "LOG_INFO"...

OlegHahm commented:
1. It's 5 lines for muting the kernel, the rest of this PR is fixing the log level enums.
2. Logging "This is RIOT" seems weird to me.

kaspar030 commented:
@OlegHahm I don't like it.

- printk is IMHO another redundant way of outputting information
- log levels as enums were nice, changing into defines just to be able to include inttypes.h (which should be done where needed) seems weird

Are you OK with closing this PR?

OlegHahm commented:
Actually no. Will explain later.

OlegHahm commented:
IMO this PR solves several issues:
 * get rid of the name collision between severity level and the macro for doing the actual logging
 * using CPP constants instead of enums in order to allow for preprocessor comparison of log levels
 * getting rid of printf() in the kernel

kaspar030 commented:
I see the name collision as issue, not the others.
I chose enums to force using of real if(), in order to get logging statements syntax checked.

Getting rid of "printf" by introducing "printk" doesn't solve anything.

OlegHahm commented:
> Getting rid of "printf" by introducing "printk" doesn't solve anything.

It's the same thing you introduced in #3888.

OlegHahm commented:
Well, at least it was here - until you complained about it.

kaspar030 commented:
> Getting rid of "printf" by introducing "printk" doesn't solve anything.
> 
> It's the same thing you introduced in #3888.

printk has a similar effect, but by introducing another output mechanism. What's wrong with using LOG_* for core's needs to print stuff?

If you don't like the name LOG for our level-based output-function - OK. We can find another. But introducing printk for that is - IMHO - an ugly workaround. If LOG_* doesn't cut it, we should improve it, not work around, don't you think?

And using defines instead of enums - that was supposed to make it possible to use the preprocessor to include inttypes.h, right? Since when do we assume that a header file needs to be included?

Did the name clash cause any problem for you?

OlegHahm commented:
> printk has a similar effect, but by introducing another output mechanism. What's wrong with using LOG_* for core's needs to print stuff?

Nothing, I introduced in https://github.com/OlegHahm/RIOT/commit/570ef93fe4ece9377099ed4203105fae56e021fb, you objected, so I changed.

> And using defines instead of enums - that was supposed to make it possible to use the preprocessor to include inttypes.h, right? Since when do we assume that a header file needs to be included?

No, the inttypes include was your suggestion in https://github.com/RIOT-OS/RIOT/pull/3506#discussion_r35623045. The change to defines is necessary to include log level dependent code in other files. But I have no strong opinion here.

> Did the name clash cause any problem for you?

To be honest, I don't remember any more if it caused other problems than in my head, but I think it's pretty ugly and should be solved anyway.

OlegHahm commented:
@kaspar030, should I revert https://github.com/OlegHahm/RIOT/commit/46ea772c6a1d5a5c29c01540ced5baf51ae6b440 then?

OlegHahm commented:
Abandoned because lack of responsiveness.
